proc_macros/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+    let FixedMapArgs { name, variants } = parse_macro_input!(input as FixedMapArgs);
+
+    let collection_type_name = format_ident!("By{name}");
+    let all_variants_collection_name = format_ident!(
+        "ALL_{}",
+        name.to_string().to_plural().to_screaming_snake_case()
+    );
+
+    let token_enum_definition = get_token_enum_definition(&name, &variants);
+    let collection_type_definition =
+        get_collection_type_definition(&collection_type_name, &variants);
+    let iter_implementations =
+        get_iter_implementations(&name, &collection_type_name, &all_variants_collection_name);
+    let index_implementations = get_index_implementations(&name, &collection_type_name);
+    let deref_implementation = get_deref_implementation(&collection_type_name, &variants);
+    let from_usize_implementation = get_from_usize_implementation(&name, &variants);
+    let all_variants_collection_definition = get_all_variants_collection_definition(
+        &name,
+        &collection_type_name,
+        &variants,
+        &all_variants_collection_name,
+    );
+    let collection_generator_macro_definition =
+        get_collection_generator_macro_definition(&collection_type_name, &variants);
+
+    quote! {
+        #token_enum_definition
+
+        #collection_type_definition
+
+        #iter_implementations
+
+        #index_implementations
+
+        #deref_implementation
+
+        #from_usize_implementation
+
+        #all_variants_collection_definition
+
+        #collection_generator_macro_definition
+    }
+    .into()
+}
+
+fn get_token_enum_definition(name: &Ident, variants: &[Ident]) -> proc_macro2::TokenStream {
+    quote! {
+        #[derive(Copy, Clone, Debug, Eq, PartialEq, clap::ValueEnum)]
+        pub enum #name {
+            #(#variants),*
+        }
+    }
+}
+
+fn get_collection_type_definition(
+    collection_type_name: &Ident,
+    variants: &[Ident],
+) -> proc_macro2::TokenStream {
+    let len = variants.len();
+    quote! {
+        #[derive(Default)]
+        pub struct #collection_type_name<T>([T; #len]);
+    }
+}
+
+fn get_iter_implementations(
+    name: &Ident,
+    collection_type_name: &Ident,
+    all_variants_collection_name: &Ident,
+) -> proc_macro2::TokenStream {
+    let iter_struct_name = format_ident!("{collection_type_name}Iter");
+    let values_struct_name = format_ident!("{collection_type_name}Values");
+
+    quote! {
+        impl<T> #collection_type_name<T> {
+            pub fn iter(&self) -> #iter_struct_name<'_, T> {
+                #iter_struct_name::new(self)
+            }
+
+            pub fn values(&self) -> #values_struct_name<'_, T> {
+                #values_struct_name::new(self)
+            }
+        }
+
+        pub struct #iter_struct_name<'collection, T> {
+            collection: &'collection #collection_type_name<T>,
+            next_index: usize,
+        }
+
+        impl<'collection, T> #iter_struct_name<'collection, T> {
```

Comment:
I never really thing about how `Iterator` is implemented on structs that exist just for that purpose.

Reply:
Right I haven't read that many implementations of them (eg of whatever the types are returned by `Vec::into_iter()` etc) but I think they just tend to be little "state machines"?